File: src/3443.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        retVal = 0

        latitude = 0
        longitude = 0
        sSize = len(s)
        for i in range(sSize):
            if s[i] == "N":
                latitude += 1
            elif s[i] == "S":
                latitude -= 1
            elif s[i] == "E":
                longitude += 1
            elif s[i] == "W":
                longitude -= 1
            retVal = max(retVal, min(abs(latitude) + abs(longitude) + k * 2, i + 1))

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for s, k in zip(["NWSE", "NSWWEW"], [1, 3]):
            # /* Example
            #  *  Input: s = "NWSE", k = 1
            #  *  Output: 3
            #  *
            #  *  Input: s = "NSWWEW", k = 3
            #  *  Output: 6
            #  */
            logging.info("Input: s = \"%s\", k = %s", s, k)

            retVal = pSolution.maxDistance(s, k)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/3443.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int maxDistance(string s, int k) {
        int retVal = 0;

        int latitude = 0;
        int longitude = 0;
        int sSize = s.size();
        for (int i = 0; i < sSize; i++) {
            switch (s[i]) {
                case 'N':
                    latitude++;
                    break;
                case 'S':
                    latitude--;
                    break;
                case 'E':
                    longitude++;
                    break;
                case 'W':
                    longitude--;
                    break;
            }
            retVal = max(retVal, min(abs(latitude) + abs(longitude) + k * 2, i + 1));
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        string s;
        int k;
    };
    vector<subject> testData{{"NWSE", 1}, {"NSWWEW", 3}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: s = "NWSE", k = 1
     *  Output: 3
     *
     *  Input: s = "NSWWEW", k = 3
     *  Output: 6
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: s = \"" << testData[i].s << "\", k = " << testData[i].k << "\n";

        answer = cSolution.maxDistance(testData[i].s, testData[i].k);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
